src/item.cpp

```diff
@@ -6393,186 +6443,225 @@ void item::update_temp( const int temp, const float insulation )
-            // increment toward cold, remove counters until we don't have any,
-            // then add warm tag and start to increment with WARM tag
-            item_counter -= loop_diff;
-            if( item_counter < 0 ) {
-                item_tags.insert( "WARM" );
-                item_counter = std::max( 600, -item_counter );
-                is_warm = true;
-            }
-        } else {
-            // no tags yet and not heating or cooling
-            item_counter -= loop_diff;
-            if( item_counter < 0 ) {
-                item_counter = 0;
-                return;
+    // If item has not been processed for two days just set it to environment temperature
+    // Or if the item has negative energy (it not been processed ever)
+    // This can cause incorrect freezing/unfreezing when the temperature is close to freezing temperature and the freezing/unfreezing would take longer than two days
+    if( to_turns<int>( time ) > 28800 || specific_energy < 0 ) {
+        set_temperature( env_temperature );
+        return;
+    }
+
+    // specific_energy = item thermal energy (mJ/g). Stored in the item
+    // temperature = item temperature (dK deci (=0.1) kelvins). Stored in the item
+    const float conductivity_term = 460 / insulation;
+    const float freezing_temperature = temp_to_kelvin( type->comestible->freeze_point );  // K
+    const float specific_heat_liquid = type->comestible->spec_heat_liquid; // J/g K
+    const float specific_heat_solid = type->comestible->spec_heat_solid; // J/g K
+    const float latent_heat = type->comestible->lat_heat; // J/kg
+    const float surface_area = pow( 0.000001 * to_milliliter( volume() ), 2 / 3 ); // m^2
+    float freeze_percentage = 0;
+
+    const float completely_frozen_specific_energy = specific_heat_solid *
+            freezing_temperature;  // Energy that the item would have if it was completely solid at freezing temperature
+    const float completely_liquid_specific_energy = completely_frozen_specific_energy +
+            latent_heat; // Energy that the item would have if it was completely liquid at freezing temperature
+
+    float new_specific_energy;
+    float new_item_temperature;
+    int extra_time;
+
+    // Temperature calculations based on Newton's law of cooling.
+    // Calculations are done assuming that the item stays in its phase.
+    // This assumption can cause over heating when transitioning from meltiong to liquid.
+    // In other transitions the item may cool/heat too little but that won't be a problem.
+    if( 0.00001 * specific_energy < completely_frozen_specific_energy ) {
+        // Was solid.
+        new_item_temperature = ( ( old_temperature - env_temperature )
+                                 * exp( - to_turns<int>( time ) * surface_area * conductivity_term / ( mass * specific_heat_solid ) )
+                                 + env_temperature );
+        new_specific_energy = new_item_temperature * specific_heat_solid;
```

Comment:
All these ``Was solid`` and ``Was liquid`` cases looks very similar to the current implementation with discrete ``warm`` and ``cold`` states. And new ``extra_time`` variable for doubtful situation should not be here.
I suppose the entire block from 6499 to 6581can be replaced with
```
new_specific_energy += ( get_energy_from_temperature( env_temperature ) - new_specific_energy ) * 
                       ( 1 -exp( - to_turns<int>( time ) * surface_area * conductivity_term / mass ) );

new_item_temperature = get_temp_from_energy( new_specific_energy );
```
It does not matter for the energy values if the item is liquid or frozen, so this equation replaces all if statements. Actual phase state will be calculated in ``get_temp_from_energy`` based if melting temperature and heat capacities.

Reply:
There are three cases because the energy flow is different.

https://i.imgur.com/5CbuvnF.png

Solid and liquid are almost identical e^kx curve. They just have different value for specific heat in them. Liquid is also offset by the latent heat.

Partially frozen item behaves completely differently. Since the temperature doesn't change the rate of energy flowing is constant.

The extra time happens often enought that it should be there. The error without it would be significant at >1h time delta. So practically every time you leave a melting item that was almost melted in your base in the morning and come back in the evening.
It prevents these two errors from happening:
https://i.imgur.com/2vgBIQB.png (heating up)
https://i.imgur.com/ozvtvk2.png (cooling down)

Reply:
> Partially frozen item behaves completely differently. Since the temperature doesn't change the rate of energy flowing is constant.

Any reference for this? It is a direct consequence from the initial heat rate equation, but this equation can not be used for freezing and melting, because it will give zero heat exchange.
I suppose this melting region is from 580 to 900 energies on your plots. And Tim is Time, not The Incredible Machine, right?

If you a using any equation - you must know its limits. As it was for using linear approximation. Otherwise you will get very large error.
And the best thing to do is to use more general equation. Something like using direct solution for differential equations instead of approximations or calculate rate based on energy as I already suggested.

Reply:
At temperatures below 0 °C water is solid.
At temperatures above 0 °C water is liquid.
So it is clear that when the water is both liquid and solid it can not be at any other temperature than 0 °C. The energy added in it go into the phase change instead of temperature increase.
(Same for boiling water)

This can even be used to calibrate thermometers. Set 0 °C to ice-water mixture and 100 0 °C to boiling water.

The rate energy change at time "t" (the linear approximation first used) is the derivative of energy at time "t". But the rate of energy change depends on temperature difference which depends on rate of energy change (with solid/liquid). (or if using equation for temperature instead of energy: the rate of temperature change depends on rate of energy change which depends on rate of temperature change).  In oter words the derivative of the function contains the function itself. So you get a first order differential equation and the solution  is some kind of E(t) = a + e^kt equation.

But the melting item does not change temperature. So the temperature difference does not depend on energy flow. So the solution is simple E(t) = a + bt.

So the full solution is a system of equations

E(t) = 
a1 + exp(k1 * t) , when below freezing temperature
a2 + exp(k2 * t) , when above freezing temperature
a3 + b * t           , when at freezing temperature

Reply:
> http://hyperphysics.phy-astr.gsu.edu/hbase/thermo/phase.html
Yes, and? I still somehow remember all this from my mid-school lessons. My point is what the heat change is different for different ice levels (different heat energies), even if the temperature stay the same 0C.

For example, we irl put plastic bottle of hot water into large reservoir with cold water at 0C and 45% ice. After few days water in the bottle will be at 0C and 45%, same as outside. This means after cooling to 0C water still will lost thermal energy, until it will have 45% of ice. 
And this heat change will be much smaller for water with 44% ice compare to water with 1% ice. So this water will change from 1% to 2% ice much faster, rather then from 43% to 44%. 

I just propose a solution which will resolve this problem. And it is much shorter in the code and easier to understand.

Reply:
Temperature gradient and insulation (+convection) are the only things that matter for the flow of energy. Different ice % levels may have slightly different insulation and convection but other than that they will behave identically (and if you mix it the differences go away).

If you warm water into ice-water mixture and there is enough ice that the warm water wont' melt it all and you mix the ice-water to stop layer of warm water forming on the surface of the bottle then the ammount of ice is irrelevant.

Also the warm water bottle into 0 °C ice-water mixture won't freeze. It will cool down until there is no temperature difference and stop. ( At 0 °C. Or earlier if the warm bottle is hot enough to melt all the ice). 
When they are both at 0 °C nothing will happen. No temperature difference -> no cooling. Maybe eventually the ice will all be on the surface (including in the bottle if it is on the surface) but I doubt "days" would be long enough for that to happen.

You may be thinking of freezing water in sub zero temperature. The surface will freeze faster than the insides. As the water freezes the convection stops effectively increasing the insulation.